File: backend/routes/ai.routes.js
import express from 'express';
import { generateRoadmap, analyzeError, processImageFeedback, chatWithAssistant, generateVisualSummary, generateMermaidDiagram } from '../services/orchestrator.service.js';
import { callBedrock, isBedrockConfigured } from '../services/bedrock.service.js';

const router = express.Router();

// 1. Roadmap Generator
router.post('/roadmap', async (req, res, next) => {
    try {
        const { goal, aiSettings } = req.body;
        if (!goal) return res.status(400).json({ error: 'Goal is required' });
        const result = await generateRoadmap(goal, aiSettings);
        res.json(result);
    } catch (error) {
        console.error('[ROADMAP]', error.message);
        res.status(500).json({ error: error.message });
    }
});

// 2. Error / Code Analysis
router.post('/feedback', async (req, res, next) => {
    try {
        const { code, question, aiSettings } = req.body;
        if (!code) return res.status(400).json({ error: 'Code/Answer is required' });
        const result = await analyzeError(code, question, aiSettings);
        res.json(result);
    } catch (error) {
        console.error('[FEEDBACK]', error.message);
        res.status(500).json({ error: error.message });
    }
});

// 3. Photo Review
router.post('/photo-review', async (req, res, next) => {
    try {
        const { imageFile, base64Image, imageType, context, aiSettings } = req.body;
        const result = await processImageFeedback(imageFile, base64Image, imageType, context, aiSettings);
        res.json(result);
    } catch (error) {
        console.error('[PHOTO-REVIEW]', error.message);
        res.status(500).json({ error: error.message });
    }
});

// 4. Chat (simple JSON, no streaming)
router.post('/chat', async (req, res, next) => {
    try {
        // AIAssistant payload is slightly different; using 'messages' instead of message/history
        const { messages, aiSettings } = req.body;
        if (!messages) return res.status(400).json({ error: 'Messages are required' });
        const result = await chatWithAssistant(messages, aiSettings);
        res.json(result);
    } catch (error) {
        console.error('[CHAT]', error.message);
        res.status(500).json({ error: error.message });
    }
});

// 5. Visual Summary Generator
router.post('/visual-summary', async (req, res, next) => {
    try {
        const { topic, aiSettings } = req.body;
        if (!topic) return res.status(400).json({ error: 'Topic is required' });
        const result = await generateVisualSummary(topic, aiSettings);
        res.json(result);
    } catch (error) {
        console.error('[VISUAL-SUMMARY]', error.message);
        res.status(500).json({ error: error.message });
    }
});

// POST /api/mermaid
router.post('/mermaid', async (req, res) => {
    try {
        const { topic, aiSettings } = req.body;
        if (!topic) {
            return res.status(400).json({ error: 'Topic is required' });
        }

        const data = await generateMermaidDiagram(topic, aiSettings);
        res.json(data);
    } catch (error) {
        console.error('Mermaid Diagram Request Error:', error);
        res.status(500).json({ error: 'Failed to generate architecture diagram' });
    }
});

// TEST endpoint — verifies live AWS Bedrock connection
router.get('/test-bedrock', async (req, res) => {
    try {
        const configured = isBedrockConfigured();
        console.log('📡 Bedrock test — configured:', configured);
        if (!configured) return res.json({ ok: false, error: 'Bedrock credentials not configured in .env' });

        const result = await callBedrock(
            process.env.BEDROCK_CHAT_MODEL_ID || 'amazon.nova-micro-v1:0',
            'You are a test assistant.',
            [{ role: 'user', content: 'Reply with exactly one word: BEDROCK_OK' }]
        );
        res.json({ ok: true, response: result });
    } catch (e) {
        console.error('[TEST-BEDROCK]', e.message);
        res.status(500).json({ ok: false, error: e.message });
    }
});

export default router;

File: synapse-fastapi/routes/ai_routes.py
from fastapi import APIRouter, HTTPException
from models.schemas import (
    RoadmapRequest, ErrorAnalysisRequest, PhotoReviewRequest,
    ChatRequest, VisualSummaryRequest, MermaidRequest, AISettings,
)
from services.orchestrator import (
    generate_roadmap, analyze_error, process_image_feedback,
    chat_with_assistant, generate_visual_summary, generate_mermaid_diagram,
)
from services.bedrock_service import call_bedrock, is_bedrock_configured, BEDROCK_MODELS
import os

router = APIRouter(prefix="/api")


# ── 1. Roadmap Generator ──────────────────────────────────────────────────────
@router.post("/roadmap")
async def roadmap(req: RoadmapRequest):
    if not req.goal:
        raise HTTPException(status_code=400, detail="goal is required")
    try:
        result = await generate_roadmap(req.goal, req.aiSettings)
        return result
    except Exception as e:
        print(f"[ROADMAP] {e}")
        raise HTTPException(status_code=500, detail=str(e))


# ── 2. Error Analysis ─────────────────────────────────────────────────────────
@router.post("/error-analysis")
async def error_analysis(req: ErrorAnalysisRequest):
    if not req.error:
        raise HTTPException(status_code=400, detail="error is required")
    try:
        result = await analyze_error(req.error, req.code, req.language or "unknown", req.aiSettings)
        return result
    except Exception as e:
        print(f"[ERROR-ANALYSIS] {e}")
        raise HTTPException(status_code=500, detail=str(e))


# ── 3. Photo Review ───────────────────────────────────────────────────────────
@router.post("/photo-review")
async def photo_review(req: PhotoReviewRequest):
    try:
        result = await process_image_feedback(
            req.imageFile, req.base64Image, req.imageType or "image/jpeg",
            req.context or "", req.aiSettings,
        )
        return result
    except Exception as e:
        print(f"[PHOTO-REVIEW] {e}")
        raise HTTPException(status_code=500, detail=str(e))


# ── 4. Chat ───────────────────────────────────────────────────────────────────
@router.post("/chat")
async def chat(req: ChatRequest):
    # Support both new format (messages[]) and old format (message + history)
    messages = []
    if req.messages:
        messages = [m.model_dump() for m in req.messages]
    elif req.message:
        history = [m.model_dump() for m in req.history] if req.history else []
        messages = history + [{"role": "user", "text": req.message, "content": req.message}]

    if not messages:
        raise HTTPException(status_code=400, detail="messages are required")

    try:
        result = await chat_with_assistant(messages, req.aiSettings)
        return result
    except Exception as e:
        print(f"[CHAT] {e}")
        raise HTTPException(status_code=500, detail=str(e))


# ── 5. Visual Summary ─────────────────────────────────────────────────────────
@router.post("/visual-summary")
async def visual_summary(req: VisualSummaryRequest):
    if not req.topic:
        raise HTTPException(status_code=400, detail="topic is required")
    try:
        result = await generate_visual_summary(req.topic, req.aiSettings)
        return result
    except Exception as e:
        print(f"[VISUAL-SUMMARY] {e}")
        raise HTTPException(status_code=500, detail=str(e))


# ── 6. Mermaid Diagram ────────────────────────────────────────────────────────
@router.post("/mermaid")
async def mermaid(req: MermaidRequest):
    if not req.topic:
        raise HTTPException(status_code=400, detail="topic is required")
    try:
        result = await generate_mermaid_diagram(req.topic, req.aiSettings)
        return result
    except Exception as e:
        print(f"[MERMAID] {e}")
        raise HTTPException(status_code=500, detail=str(e))


# ── 7. Bedrock Connectivity Test ──────────────────────────────────────────────
@router.get("/test-bedrock")
async def test_bedrock():
    try:
        configured = is_bedrock_configured()
        if not configured:
            return {"ok": False, "error": "Bedrock credentials not configured in .env"}

        result = call_bedrock(
            os.getenv("BEDROCK_CHAT_MODEL_ID", "amazon.nova-micro-v1:0"),
            "You are a test assistant.",
            [{"role": "user", "content": "Reply with exactly one word: BEDROCK_OK"}],
        )
        return {"ok": True, "response": result}
    except Exception as e:
        print(f"[TEST-BEDROCK] {e}")
        raise HTTPException(status_code=500, detail=str(e))
